azure/src/main/java/org/apache/iceberg/azure/blob/AzureBlobInputStream.java

```diff
@@ -0,0 +1,130 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class AzureBlobInputStream extends SeekableInputStream {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AzureBlobInputStream.class);
+  private static final Joiner TRACE_JOINER = Joiner.on("\n\t");
+
+  private final StackTraceElement[] createStack;
+  private final AzureURI azureURI;
+  private final AzureProperties azureProperties;
+  private final BlobClient blobClient;
+
+  private long pos = 0L;
+  private boolean closed = false;
+  private BlobInputStream stream;
+
+  public AzureBlobInputStream(AzureURI azureURI, AzureProperties azureProperties, BlobClient blobClient) {
+    this.createStack = Thread.currentThread().getStackTrace();
+    this.azureURI = azureURI;
+    this.azureProperties = azureProperties;
+    this.blobClient = blobClient;
+    openStream(/* offset = */ 0);
+  }
+
+  @Override
+  public long getPos() {
+    return pos;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "Stream already closed.");
+    Preconditions.checkArgument(newPos >= 0, "New position cannot be negative: %s.", newPos);
+
+    if (newPos == pos) {
+      // Already at the specified position.
+      return;
+    }
+
+    if (newPos > pos) {
+      // Seeking forward.
+      final long bytesToSkip = newPos - pos;
+      // BlobInputStream#skip only repositions the internal pointers,
+      // the actual bytes are skipped when BlobInputStream#read is invoked.
+      final long bytesSkipped = stream.skip(bytesToSkip);
+    } else {
+      // Seeking backward.
+      stream.close();
+      openStream(newPos);
+    }
+
+    pos = newPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    Preconditions.checkState(!closed, "Cannot read: stream already closed.");
+    pos++;
+    return stream.read();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "Cannot read: stream already closed.");
+    int numOfBytesRead = stream.read(b, off, len);
+    pos += numOfBytesRead;
+    return numOfBytesRead;
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (closed) {
+      return;
+    }
+
+    super.close();
+    stream.close();
+    closed = true;
+  }
+
+  @SuppressWarnings("checkstyle:NoFinalizer")
+  @Override
+  protected void finalize() throws Throwable {
+    super.finalize();
+    if (!closed) {
+      close(); // releasing resources is more important than printing the warning
+      String trace = TRACE_JOINER.join(Arrays.copyOfRange(createStack, 1, createStack.length));
+      LOG.warn("Unclosed input stream created by:\n\t{}", trace);
+    }
+  }
+
+  private void openStream(long offset) {
+    final BlobInputStreamOptions options = new BlobInputStreamOptions().setRange(new BlobRange(offset))
+        .setBlockSize(azureProperties.readBlockSize(azureURI.storageAccount()));
```

Comment:
We usually put each chained method on a separate line.

Reply:
Ack